libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -122,78 +152,109 @@ impl Replacements {
         Some(replacements)
     }
 
-    fn insert(&mut self, k: ExprPat, v: Expr) -> Option<Expr> {
-        self.map.insert(k, v)
+    fn insert(&mut self, k: &Rc<ExprPat>, v: Rc<Expr>) -> Option<Rc<Expr>> {
+        self.map.insert(hash(k.as_ref()), v)
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::{parse_expression, parse_expression_pattern, scan};
+
+    fn parse_rule(prog: &str) -> ExprPat {
+        let (expr, _) = parse_expression_pattern(scan(prog));
+        expr
+    }
+
```

Comment:
can't we just return parse_expression_pattern.unwrap()?

Reply:
No, `parse_expression_pattern` returns a tuple which we have to destruct here.